### PR TITLE
finished tic_tac_toe

### PR DESCRIPTION
**Pull request by DanielMitiku**

### CODE REVIEW

**Review by rvvergara — CHANGES_REQUESTED** [review]
Hi Daniel,

Good effort you put in for this initial version of the project. However, it appears that the project didn't fully satisfy the OOP requirement (aside from the fact that you created a class called TicTacToe). Here are my thoughts:

1. I notice that it was only you who worked on this but the project is supposed to be collaborative. Correct me if I'm wrong.
2. To build a truly Object Oriented application you need to break down your code into several objects/classes that are reusable/independent of one another. Here are some things to consider:
 - The three classes you will need to build are:
 > Board class -> all things that pertain to the board
> Player class -> all things that pertain to players (it should contain things like how a player chooses a square in the board, etc)
> Game class -> basically controls the flow of the game (which includes making sure each player makes successive turns until somebody has won or there are no more empty squares in the board)

3. Make sure you don't mix any User Interface related logic into your classes. I suggest you created modules that you will just include (if absolutely necessary) in your classes. 

Let's start with this and submit another code review request when you've built all three classes. I suggest for now don't focus on the User Interface but just on the classes

File: lib/player.rb [hunk 1]
```diff
@@ -0,0 +1,15 @@
+class Player
```
Comment (rvvergara) [inline lib/player.rb:1]: Can you run me down on how you came up with the implementation of the Player class? The way I see it it's not a class for a particular player object but more of a class that interchanges the two players. Can we make it simpler than this? The Player object isn't supposed to be responsible for interchanging "player turns". The player, in this regard, should only be responsible for one thing - making moves (ergo choosing squares in the tictactoe board).

File: lib/game.rb [hunk 2]
```diff
@@ -0,0 +1,73 @@
+class Game
```
Comment (rvvergara) [inline lib/game.rb:3]: Good job breaking down the game objects into three classes. So our work begins. Ok here's the next thing we need to do. We need to answer the following questions:

1. The Game's main responsibility is to control the flow of the game. Meaning making sure that each player takes his/her turn while nobody has won or there are still empty spaces in the board - what are the methods that we need to do this? If certain methods don't fall within this responsibility then they don't belong to the Game class.
2. Among the methods we will place here, which ones do we need to be exposed as a public interface of the Game class? In other words, which methods do we need outside code to use for game? For example, do we need the `game.move` method to be called by an outside code? For methods we need to be public interfaces we can put in the `attr_reader` line. But methods that aren't needed as public interface we should put them under the `private` line.
I look into the Game class and don't see any way (as you didn't put any attributes as `attr_reader`) outside code (for instance, the main.rb file) to use the Game class.

File: lib/game.rb [hunk 3]
```diff
@@ -0,0 +1,73 @@
...
+            [0,4,8], # left_diagonal 
+            [6,4,2] # right_diagonal 
+            ]
+    end
+
+    def move(index, player)
+        @game_board.board[index] = player
+    end
+
+    def valid_move?(index)
```
Comment (rvvergara) [inline lib/game.rb]: Outside of the Game class did we need to call `valid_move?`? If not, then we better place this method as a `private` method

File: lib/game.rb [hunk 4]
```diff
@@ -0,0 +1,73 @@
...
+        if index.between?(0,8) && !@game_board.position_taken?(index)
+           return true
+        end
+        return false
+    end
+
+    def full?
```
Comment (rvvergara) [inline lib/game.rb:49]: Same comment as the `valid_move?` method

File: lib/game.rb [hunk 5]
```diff
@@ -0,0 +1,73 @@
...
+        @game_board.board.all? {|i| i == "X" || i == "O"}
+    end
+
+    def won?
```
Comment (rvvergara) [inline lib/game.rb:53]: Same comment as the `valid_move?` method.

Also is there a way we can make this method simpler? You can use array methods like `all?` and `any?` to make this simpler and write fewer lines.

File: lib/game.rb [hunk 6]
```diff
@@ -0,0 +1,73 @@
...
+            [6,7,8], # bottom_row 
+            [0,3,6], # left_column 
+            [1,4,7], # center_column 
+            [2,5,8], # right_column 
+            [0,4,8], # left_diagonal 
+            [6,4,2] # right_diagonal 
+            ]
+    end
+
+    def move(index, player)
```
Comment (rvvergara) [inline lib/game.rb]: This is probably the best place to insert the logic of interchanging between players. As mentioned, the Game class's sole responsibility is to ensure each player takes a turn alternately while nobody has won or there are still empty squares.

File: lib/game.rb [hunk 7]
```diff
@@ -0,0 +1,73 @@
+class Game
+    def initialize(game_board, game_player)
+        @game_board = game_board
+        @game_player = game_player
```
Comment (rvvergara) [inline lib/game.rb]: Does the tictactoe game only take in one player? I know that in your Player class you have put the logic of changing the current_player. And that's where the main confusion will be in here. Looking into this code one would think that the Game class only needs one player (when in fact, unless you have a one player mode which a human player can play against the computer, you need two players for the game). So we really need to change the implementation of Player class to erase this confusion

File: lib/game.rb [hunk 8]
```diff
@@ -0,0 +1,73 @@
...
+          position_3 = @game_board.board[win_index_3] # value of game_board.board at win_index_3
+          if (position_1 == position_2 && position_2 == position_3 && @game_board.position_taken?(win_index_1))
+            @winner = position_1
+            return true 
+          end
+        end
+        return false
+    end
+
+    def draw?
```
Comment (rvvergara) [inline lib/game.rb:69]: I know what this `draw?` method is doing. It is looking if we have a draw. However, how this method can be simpler. Ok, question, in a playing tictactoe in a piece of paper, how would we tell if it was a draw?

File: lib/game.rb [hunk 9]
```diff
@@ -0,0 +1,73 @@
...
+    end
+
+    def over?
+        if draw? || won? || full?
+          return true
+        end
+        false
+    end
+    
+    def winner
```
Comment (rvvergara) [inline lib/game.rb]: The `won?` method I suppose is testing if a player is the winner right? Then why do we need to have an extra method for determining the `winner` here?

**Review by rvvergara — CHANGES_REQUESTED** [review]
Hi @DanielMitiku,

Thanks for the clarification regarding why you were the only one who worked into this project. I hope everything's well with you my friend.

Anyway, I'd like to commend you then on the effort you're putting into this project as this is a tough one to do without a partner and yet you're hustling. For that, my hats off to you.

Now, we have some work to do to make this application truly OOP. Please see my individual comments. I haven't touched on the main.rb file yet (and we have lots of work to do there) because I first want you to focus on the three classes (Game, Board and Player) as they still need a lot of improvement.

**Comment by DanielMitiku** [comment]
Hey @rvvergara, it is currently impossible to do apply all the comments you suggested. It would take forever because you haven't even seen the main.rb file. I would suggest you clearly say what functions need to be implemented in each class. That would be easier. Waiting for your reply

**Comment by rvvergara** [comment]
Hi @DanielMitiku. In regards to the main.rb file. It's doing too much. If it's the main file of the application I would probably place only the ff stuff:
- initialization of player1 and player2
- initialization of board
- initialization of game
- running a method in the game that cycles between player turns until the game is finished

```
player_1 = Player.new(<arguments needed to create player>)
player_2 = Player.new(<arguments needed to create player>)
board = Board.new(<arguments needed to create board>)
game = Game.new(<arguments needed to create game>)
game.some_method #main method inside game that cycles between players until one has won or it's a draw
```

File: lib/game.rb [hunk 10]
```diff
@@ -0,0 +1,123 @@
+module Display
```
Comment (rvvergara) [inline lib/game.rb]: Why don't we separate this module into another file?

File: bin/main.rb [hunk 11]
```diff
@@ -0,0 +1,34 @@
+require '../lib/board.rb'
+require '../lib/game.rb'
+require '../lib/player.rb'
+
+def choose_player
```
Comment (rvvergara) [inline bin/main.rb]: We can just place this in another file. As I mentioned in my previous review the main.rb file is best only populated with the object initializations and running the main game method.

**Review by rvvergara — CHANGES_REQUESTED** [review]
Hi @DanielMitiku ,

Just two minor changes (just transferring a module and a method into a separate file, pls see respective comments). 

Pls work on these changes and let's have one last round of reviews (I promise that'll be the last because after that I will just leave you with some points to consider).

**Review by rvvergara — APPROVED** [review]
Hi @DanielMitiku ,

Great! Now as promised I won't be tying you up further on this project. While of course there are lots of improvements that can be made I believe you have satisfied the minimum requirements. For that

# :+1: 

You may now complete this project and merge to master.

Now here are some things you should remember when you are using OOP (not just for Ruby projects but later on when you get to JavaScript):

# SOLID (read more about it online)
**S**ingle Responsibility Principle
**O**pen for extension and closed for modification
**L**iskov substitutionary principle
**I**nterface segregation
**D**ependency Injection

This is one of the most powerful concepts there is in software engineering. It lets us build code that are not just working, but also efficient, highly maintainable and collaboration and teamwork friendly.

As you perhaps (I hope) learned from our series of code reviews, I would normally request you to separate code and have your classes only do one thing (its sole responsibility). The reason is that you want your code to be reusable and extendable. Currently your game only features a 2-player mode. What if later on (in a real similar project) your client or boss requires you to build a 1 player feature? The ease at which you are able to extend your app with this new feature is dependent on how far you went in making it truly OOP and SOLID.

**Comment by DanielMitiku** [comment]
@rvvergara

Thank you!

**Pull request merged**